Index: qtbase_temp/src/widgets/widgets/qlineedit_p.cpp
===================================================================
--- qtbase_temp.orig/src/widgets/widgets/qlineedit_p.cpp
+++ qtbase_temp/src/widgets/widgets/qlineedit_p.cpp
@@ -489,9 +489,24 @@ QLineEditPrivate::SideWidgetParameters Q
     Q_Q(const QLineEdit);
     SideWidgetParameters result;
     result.iconSize = q->style()->pixelMetric(QStyle::PM_SmallIconSize, nullptr, q);
-    result.margin = result.iconSize / 4;
-    result.widgetWidth = result.iconSize + 6;
-    result.widgetHeight = result.iconSize + 2;
+    
+    bool ok = true;
+    int margin = q->property("_d_dtk_lineeditActionMargin").toInt(&ok);
+
+    if (!ok)
+        margin = result.iconSize / 4;
+    int widgetWidth = q->property("_d_dtk_lineeditActionWidth").toInt(&ok);
+
+    if (!ok)
+        widgetWidth = 6;
+    int widgetHeight = q->property("_d_dtk_lineeditActionHeight").toInt(&ok);
+
+    if (!ok)
+        widgetHeight = 2;
+
+    result.margin = margin;
+    result.widgetWidth = result.iconSize + widgetWidth;
+    result.widgetHeight = result.iconSize + widgetHeight;
     return result;
 }
 
