Author: zhanghao <zhanghao@uniontech.com>
Date:   Wed Nov 17 09:27:09 2021 +0800
Subject: QFontComboBox don't response qApp fontDatabaseChanged()
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/380720
---

Index: qtbase-opensource-src/src/widgets/widgets/qfontcombobox.cpp
===================================================================
--- qtbase-opensource-src.orig/src/widgets/widgets/qfontcombobox.cpp
+++ qtbase-opensource-src/src/widgets/widgets/qfontcombobox.cpp
@@ -309,6 +309,10 @@ public:
 void QFontComboBoxPrivate::_q_updateModel()
 {
     Q_Q(QFontComboBox);
+
+ if (QCoreApplication::closingDown())
+        return;
+
     const int scalableMask = (QFontComboBox::ScalableFonts | QFontComboBox::NonScalableFonts);
     const int spacingMask = (QFontComboBox::ProportionalFonts | QFontComboBox::MonospacedFonts);
 
Index: qtbase-opensource-src/tests/auto/widgets/widgets/qfontcombobox/tst_qfontcombobox.cpp
===================================================================
--- qtbase-opensource-src.orig/tests/auto/widgets/widgets/qfontcombobox/tst_qfontcombobox.cpp
+++ qtbase-opensource-src/tests/auto/widgets/widgets/qfontcombobox/tst_qfontcombobox.cpp
@@ -253,15 +253,18 @@ void tst_QFontComboBox::writingSystem()
 // protected void currentFontChanged(QFont const& f)
 void tst_QFontComboBox::currentFontChanged()
 {
-    SubQFontComboBox box;
-    QSignalSpy spy0(&box, SIGNAL(currentFontChanged(QFont)));
+   // The absence of this file does not affect the test results
+    QFontDatabase::addApplicationFont("ArianaVioleta-dz2K.ttf");
 
-    if (box.model()->rowCount() > 2) {
-        QTest::keyPress(&box, Qt::Key_Down);
+    SubQFontComboBox *box = new SubQFontComboBox;
+    QSignalSpy spy0(box, SIGNAL(currentFontChanged(QFont)));
+
+    if (box->model()->rowCount() > 2) {
+        QTest::keyPress(box, Qt::Key_Down);
         QCOMPARE(spy0.count(), 1);
 
         QFont f( "Sans Serif" );
-        box.setCurrentFont(f);
+        box->setCurrentFont(f);
         QCOMPARE(spy0.count(), 2);
     } else
         qWarning("Not enough fonts installed on test system. Consider adding some");
