Index: qtbase-opensource-src/src/widgets/widgets/qwidgettextcontrol.cpp
===================================================================
--- qtbase-opensource-src.orig/src/widgets/widgets/qwidgettextcontrol.cpp
+++ qtbase-opensource-src/src/widgets/widgets/qwidgettextcontrol.cpp
@@ -1942,10 +1942,13 @@ void QWidgetTextControlPrivate::contextM
     if (!menu)
         return;
     menu->setAttribute(Qt::WA_DeleteOnClose);
-    if (auto *window = static_cast<QWidget *>(parent)->window()->windowHandle()) {
-        QMenuPrivate::get(menu)->topData()->initialScreenIndex =
+
+    if (auto *widget = qobject_cast<QWidget *>(parent)) {
+        if (auto *window = widget->window()->windowHandle())
+            QMenuPrivate::get(menu)->topData()->initialScreenIndex =
                 QGuiApplication::screens().indexOf(window->screen());
     }
+
     menu->popup(screenPos);
 #endif
 }
