Author: Tang Haixiang <tanghaixiang@uniontech.com>
Date:   Wed May 11 14:09:12 2022 +0800
Subject: QHighDpi: revert window-on-multiple-screens handling
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/300513

---

Index: qtbase-opensource-src_set0/src/gui/kernel/qhighdpiscaling_p.h
===================================================================
--- qtbase-opensource-src_set0.orig/src/gui/kernel/qhighdpiscaling_p.h
+++ qtbase-opensource-src_set0/src/gui/kernel/qhighdpiscaling_p.h
@@ -218,8 +218,7 @@ inline QPoint position(QRectF rect) { re
 template <typename T, typename C>
 T fromNativePixels(const T &value, const C *context)
 {
-    QPoint nativePosition = position(value);
-    QHighDpiScaling::ScaleAndOrigin so = QHighDpiScaling::scaleAndOrigin(context, &nativePosition);
+    QHighDpiScaling::ScaleAndOrigin so = QHighDpiScaling::scaleAndOrigin(context);
     return scale(value, qreal(1) / so.factor, so.origin);
 }
 
