Author: Duan Ting <duanting@uniontech.com>
Date:   Thu June 20 16:30:18 2022 +0800
Subject: Ignore window icon in XCB when its size exceeds maximum request length
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/408200
---

Index: qtbase-opensource-src/src/plugins/platforms/xcb/qxcbwindow.cpp
===================================================================
--- qtbase-opensource-src.orig/src/plugins/platforms/xcb/qxcbwindow.cpp
+++ qtbase-opensource-src/src/plugins/platforms/xcb/qxcbwindow.cpp
@@ -1313,6 +1313,12 @@ void QXcbWindow::setWindowIcon(const QIc
     }
 
     if (!icon_data.isEmpty()) {
+        // Ignore icon exceeding maximum xcb request length
+        if (icon_data.size() > xcb_get_maximum_request_length(xcb_connection())) {
+            qWarning("Ignoring window icon: Size %llu exceeds maximum xcb request length %u.",
+                     icon_data.size(), xcb_get_maximum_request_length(xcb_connection()));
+            return;
+        }
         xcb_change_property(xcb_connection(),
                             XCB_PROP_MODE_REPLACE,
                             m_window,
