Author: Zhang Hao <zhanghao@uniontech.com>
Date:   Tue Jun 22 09:49:37 2021 +0800
Subject: Check scrollbar ScrollBarOverlap when computing QListView margins
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/353094
---

Index: qtbase-opensource-src/src/widgets/itemviews/qlistview.cpp
===================================================================
--- qtbase-opensource-src.orig/src/widgets/itemviews/qlistview.cpp
+++ qtbase-opensource-src/src/widgets/itemviews/qlistview.cpp
@@ -1821,6 +1821,7 @@ void QListViewPrivate::prepareItemsLayou
     // Qt::ScrollBarAlwaysOn but scrollbar extent must be deduced if policy
     // is Qt::ScrollBarAsNeeded
     int verticalMargin = (vbarpolicy == Qt::ScrollBarAsNeeded) && (flow == QListView::LeftToRight || vbar->isVisible())
+        && !q->style()->pixelMetric(QStyle::PM_ScrollView_ScrollBarOverlap, nullptr, vbar)
         ? q->style()->pixelMetric(QStyle::PM_ScrollBarExtent, nullptr, vbar) + frameAroundContents
         : 0;
     int horizontalMargin =  hbarpolicy==Qt::ScrollBarAsNeeded
Index: qtbase-opensource-src/tests/auto/widgets/itemviews/qlistview/tst_qlistview.cpp
===================================================================
--- qtbase-opensource-src.orig/tests/auto/widgets/itemviews/qlistview/tst_qlistview.cpp
+++ qtbase-opensource-src/tests/auto/widgets/itemviews/qlistview/tst_qlistview.cpp
@@ -37,6 +37,7 @@
 #include <QTest>
 #include <QTimer>
 #include <QtMath>
+#include <QProxyStyle>
 
 #include <QtTest/private/qtesthelpers_p.h>
 #include <QtWidgets/private/qlistview_p.h>
@@ -171,6 +172,7 @@ private slots:
     void itemAlignment();
     void internalDragDropMove_data();
     void internalDragDropMove();
+    void spacingWithWordWrap_data();
     void spacingWithWordWrap();
     void scrollOnRemove_data();
     void scrollOnRemove();
@@ -2796,11 +2798,40 @@ void tst_QListView::internalDragDropMove
     }
 }
 
+void tst_QListView::spacingWithWordWrap_data()
+{
+    QTest::addColumn<bool>("scrollBarOverlap");
+
+    QTest::addRow("Without overlap") << false;
+    QTest::addRow("With overlap") << true;
+}
+
 /*!
     Verify fix for QTBUG-92366
 */
 void tst_QListView::spacingWithWordWrap()
 {
+    QFETCH(bool, scrollBarOverlap);
+
+    class MyStyle : public QProxyStyle
+    {
+        bool scrollBarOverlap;
+    public:
+        MyStyle(bool scrollBarOverlap) : scrollBarOverlap(scrollBarOverlap) {}
+
+        int pixelMetric(PixelMetric metric, const QStyleOption *option = nullptr,
+                        const QWidget *widget = nullptr) const override{
+            switch (metric) {
+                case QStyle::PM_ScrollView_ScrollBarOverlap: return scrollBarOverlap;
+                default:
+                    break;
+                }
+                return QProxyStyle::pixelMetric(metric, option, widget);
+        }
+    };
+
+    QApplication::setStyle(new MyStyle(scrollBarOverlap));
+
     const int listViewResizeCount = 200;
     QWidget window;
     window.resize(300, 200);
