From: Tang Haixiang <tanghaixiang@uniontech.com>
Date: Fri Mar 5 16:08:13 2021 +0800
Subject: Gif decoder fix read error caused by ub check
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/314453
---

Index: qtbase-opensource-src/src/plugins/imageformats/gif/qgifhandler.cpp
===================================================================
--- qtbase-opensource-src.orig/src/plugins/imageformats/gif/qgifhandler.cpp
+++ qtbase-opensource-src/src/plugins/imageformats/gif/qgifhandler.cpp
@@ -490,12 +490,15 @@ int QGIFFormat::decode(QImage *image, co
             break;
           case ImageDataBlock:
             count++;
-            if (bitcount < 0 || bitcount > 31) {
-                state = Error;
-                return -1;
+            if (bitcount != -32768) {
+                if (bitcount < 0 || bitcount > 31) {
+                    state = Error;
+                    return -1;
+                }
+                accum |= (ch << bitcount);
+                bitcount += 8;
             }
-            accum|=(ch<<bitcount);
-            bitcount+=8;
+
             while (bitcount>=code_size && state==ImageDataBlock) {
                 int code=accum&((1<<code_size)-1);
                 bitcount-=code_size;
