Index: qtbase_temp/src/widgets/widgets/qlineedit.cpp
===================================================================
--- qtbase_temp.orig/src/widgets/widgets/qlineedit.cpp
+++ qtbase_temp/src/widgets/widgets/qlineedit.cpp
@@ -1946,6 +1946,7 @@ void QLineEdit::focusInEvent(QFocusEvent
             selectAll();
     } else if (e->reason() == Qt::MouseFocusReason) {
         d->clickCausedFocus = 1;
+        updateMicroFocus();
     }
 #ifdef QT_KEYPAD_NAVIGATION
     if (!QApplicationPrivate::keypadNavigationEnabled() || (hasEditFocus() && ( e->reason() == Qt::PopupFocusReason))) {
