Author: Tang Haixiang <tanghaixiang@uniontech.com>
Date:   Thu Jun 16 11:09:12 2022 +0800
Subject: QMenu: Adjust the position of the submenu
Upstream: https://codereview.qt-project.org/c/qt/qtbase/+/414728

---

Index: qtbase-opensource-src/src/widgets/widgets/qmenu.cpp
===================================================================
--- qtbase-opensource-src.orig/src/widgets/widgets/qmenu.cpp
+++ qtbase-opensource-src/src/widgets/widgets/qmenu.cpp
@@ -3664,7 +3664,7 @@ void QMenu::internalDelayedPopup()
     const QRect actionRect(d->actionRect(d->currentAction));
     QPoint subMenuPos(mapToGlobal(QPoint(actionRect.right() + subMenuOffset + 1, actionRect.top())));
     if (subMenuPos.x() > screen.right())
-        subMenuPos.setX(QCursor::pos().x());
+        subMenuPos.setX(geometry().left());
 
     const auto &subMenuActions = d->activeMenu->actions();
     if (!subMenuActions.isEmpty()) {
